File: 백준/Gold/1339. 단어 수학/단어 수학.py
n = int(input())

arr=[]
for _ in range(n):
    a = input()
    arr.append(a)

dict = {}

for word in arr:
    cnt = len(word)-1
    for w in word:
        if w not in dict:
            dict[w] = pow(10,cnt)
        else:
            dict[w] += pow(10,cnt)
        cnt -=1
dict = sorted(dict.values(), reverse=True)

res = 0
num = 9

for k in dict:
    res += k*num
    num-=1

print(res)

File: 백준/Gold/1339. 단어 수학/단어 수학.java
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine());
        String[] arr = new String[n];

        for (int i = 0; i < n; i++) {
            arr[i] = reader.readLine();
        }
        HashMap<Character,Integer> map = new HashMap<>();
        for(String word: arr){
            int len = word.length();
            for(int i=0; i<len;i++){
                char c = word.charAt(i);
                map.put(c, map.getOrDefault(c,0)+ (int) Math.pow(10,len-i-1));
            }
        }
        List<Integer> weight = new ArrayList<>(map.values());
        weight.sort((a,b)->(b-a));
        int p = 9;
        int answer = 0;
        for(int w: weight){
            answer += p*w;
            p--;
        }
        System.out.println(answer);
    }
}